pkg/event/registry_test.go

```diff
@@ -0,0 +1,108 @@
+			registry = event.NewRegistry("test-registry", npGenericKubeproxyIptables)
+			handler1 = NewTestHandler("handler1", npGenericKubeproxyIptables)
+			handler2 = NewTestHandler("handler2", npGenericKubeproxyIptables)
+			err := registry.AddHandlers(logger.NewHandler(), handler1, handler2)
+			Expect(err).NotTo(HaveOccurred())
+			cleanupTestEvents()
+		})
+
+		It("should initialize all the handlers", func() {
+			Expect(handler1.Initialized).To(BeTrue())
+			Expect(handler2.Initialized).To(BeTrue())
+		})
+
+		It("Should be called in registration order", func() {
+			_ = registry.TransitionToGateway()
+			Expect(allTestEvents).Should(Receive(Equal(TestEvent{Name: evTransitionToGateway, Handler: handler1.Name})))
+			Expect(allTestEvents).Should(Receive(Equal(TestEvent{Name: evTransitionToGateway, Handler: handler2.Name})))
+		})
+
+		It("Should work for all calls", func() {
+			endpoint := &submV1.Endpoint{ObjectMeta: v1meta.ObjectMeta{Name: "endpoint1"}}
+			node := &k8sV1.Node{ObjectMeta: v1meta.ObjectMeta{Name: "node1"}}
+
+			events := map[TestEvent]func() error{
+				{Name: evStop, Parameter: false}:                     func() error { return registry.StopHandlers(false) },
+				{Name: evTransitionToGateway}:                        registry.TransitionToGateway,
+				{Name: evTransitionToNonGateway}:                     registry.TransitionToNonGateway,
+				{Name: evNodeCreated, Parameter: node}:               func() error { return registry.NodeCreated(node) },
+				{Name: evNodeUpdated, Parameter: node}:               func() error { return registry.NodeUpdated(node) },
+				{Name: evNodeRemoved, Parameter: node}:               func() error { return registry.NodeRemoved(node) },
+				{Name: evLocalEndpointCreated, Parameter: endpoint}:  func() error { return registry.LocalEndpointCreated(endpoint) },
+				{Name: evLocalEndpointUpdated, Parameter: endpoint}:  func() error { return registry.LocalEndpointUpdated(endpoint) },
+				{Name: evLocalEndpointRemoved, Parameter: endpoint}:  func() error { return registry.LocalEndpointRemoved(endpoint) },
+				{Name: evRemoteEndpointCreated, Parameter: endpoint}: func() error { return registry.RemoteEndpointCreated(endpoint) },
+				{Name: evRemoteEndpointUpdated, Parameter: endpoint}: func() error { return registry.RemoteEndpointUpdated(endpoint) },
+				{Name: evRemoteEndpointRemoved, Parameter: endpoint}: func() error { return registry.RemoteEndpointRemoved(endpoint) },
+			}
+
+			for ev, f := range events {
+				err := f()
+				Expect(err).To(Succeed())
+				Expect(handler1.Events).Should(Receive(Equal(ev)))
+				Expect(handler2.Events).Should(Receive(Equal(ev)))
+			}
+		})
+
+		It("one handler returns error subsequent handlers should still execute", func() {
+			handler1.FailOnEvent = errors.New("I shall fail!")
+			err := registry.TransitionToGateway()
+			Expect(handler1.Events).ShouldNot(Receive())
+			Expect(handler2.Events).Should(Receive())
+			Expect(err).Should(HaveOccurred())
```

Comment:
Seeing the error handling usage now makes me wonder of the registry should return an error at all or just log if a handler fails. One handler could succeed while another fails - the caller has no knowledge of that. How is a caller to handle the returned error other than to log it...

Reply:
The events will be coming through a workqueue, we should expect that the handlers are idempotent, and that the events will be retried if any handler fails,

Probably the controller will need to put create/update/delete events in a single workqueue for something like that to work, but that's implementation we can discuss on next PRs.

Let's see how it unfolds/looks on next PRs. It would be equivalent to what we are doing today, we are just changing the shape into something more modular where code will be better organized than today.